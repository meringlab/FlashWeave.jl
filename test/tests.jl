using FlashWeave
using FlashWeave: TestResult
using Test
using DelimitedFiles

data = Matrix{Float64}(readdlm(joinpath("data", "HMP_SRA_gut", "HMP_SRA_gut_small.tsv"), '\t')[2:end, 2:end])
data_clr, mask = FlashWeave.preprocess_data_default(data, "fz", verbose=false, prec=64)
data_clr_nz, mask = FlashWeave.preprocess_data_default(data, "fz_nz", verbose=false, prec=64)
data_bin, mask = FlashWeave.preprocess_data_default(data, "mi", verbose=false, prec=64)
data_mi_nz, mask = FlashWeave.preprocess_data_default(data, "mi_nz", verbose=false, prec=64)

function compare_test_results(r1::FlashWeave.TestResult, r2::FlashWeave.TestResult)
    isapprox(r1.stat, r2.stat, rtol=1e-2) && isapprox(r1.pval, r2.pval, rtol=1e-2) && r1.df == r2.df && r1.suff_power == r2.suff_power
end

function read_exp_dict(exp_path)
    exp_mat, exp_header = readdlm(exp_path, header=true)
    exp_dict = Dict()
    for i in 1:size(exp_mat, 1)
        row = exp_mat[i, :]
        key = row[1]
        test_info = row[2:end]

        t_res = FlashWeave.TestResult(test_info...)
        if startswith(key, "exp_cond")
            exp_dict[key] = t_res
        else
            if !haskey(exp_dict, key)
                exp_dict[key] = [t_res]
            else
                push!(exp_dict[key], t_res)
            end
        end
    end
    exp_dict
end


exp_dict = read_exp_dict(joinpath("data", "tests_expected.tsv"))

for (test_name, data_norm) in [("mi", data_bin), ("mi_nz", data_mi_nz),
                               ("fz", data_clr), ("fz_nz", data_clr_nz)]
    @testset "$test_name" begin
        for cond_mode in ["uni", "condZ1", "condZ3"]

            @testset "$cond_mode" begin

                exp_res = exp_dict["exp_$(cond_mode)_$(test_name)"]

                if test_name == "fz_nz"
                    if cond_mode == "uni"
                        sub_data = @view data_norm[data_norm[:, 1] .!= 0, :]
                    else
                        sub_data = @view data_norm[(data_norm[:, 31] .!= 0) .& (data_norm[:, 21] .!= 0), :]
                    end
                else
                    sub_data = data_norm
                end

                if cond_mode == "uni"
                    pred_res = @inferred FlashWeave.test(1, collect(2:50), sub_data, test_name)
                    @test all([compare_test_results(r1, r2) for (r1, r2) in zip(pred_res, exp_res)])
                elseif cond_mode == "condZ1"
                    pred_res = @inferred FlashWeave.test(31, 21, (7,), sub_data, test_name)
                    @test compare_test_results(pred_res, exp_res)
                elseif cond_mode == "condZ3"
                    pred_res = @inferred FlashWeave.test(31, 21, (7, 14, 18,), sub_data, test_name)
                    @test compare_test_results(pred_res, exp_res)
                end

            end
        end
    end
end


# reproduce test results
# exp_dict = Dict{String, Any}()
# for (test_name, data_norm) in [("mi", data_bin), ("mi_nz", data_mi_nz),
#                                ("fz", data_clr), ("fz_nz", data_clr_nz)]
#     for cond_mode in ["uni", "condZ1", "condZ3"]
#         if test_name == "fz_nz"
#             if cond_mode == "uni"
#                 sub_data = @view data_norm[data_norm[:, 1] .!= 0, :]
#             else
#                 sub_data = @view data_norm[(data_norm[:, 31] .!= 0) .& (data_norm[:, 21] .!= 0), :]
#             end
#         else
#             sub_data = data_norm
#         end
#
#         if cond_mode == "uni"
#             test_res = FlashWeave.test(1, collect(2:50), sub_data, test_name)
#         elseif cond_mode == "condZ1"
#             test_res = FlashWeave.test(31, 21, (7,), sub_data, test_name)
#         elseif cond_mode == "condZ3"
#             test_res = FlashWeave.test(31, 21, (7, 14, 18,), sub_data, test_name)
#         end
#         exp_dict["exp_$(cond_mode)_$(test_name)"] = test_res
#     end
# end
# save(joinpath("data", "tests_expected.jld2"), exp_dict)
