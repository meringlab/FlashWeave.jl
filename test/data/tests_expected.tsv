key	stat	pval	df	suff_power
exp_condZ1_mi	-0.015573830513972549	0.004568614203379179	2	true
exp_condZ1_mi_nz	0.0	1.0	0	false
exp_condZ3_fz_nz	-0.09826893304219081	0.5278669773924074	0	true
exp_condZ3_fz	-0.07643814205965811	0.1560678789975989	0	true
exp_condZ1_fz_nz	-0.07612106982252846	0.6252967262612277	0	true
exp_uni_mi	-0.009867365175545666	0.008972880794345217	1	true
exp_uni_mi	-0.0012311609336842274	0.35599750378083306	1	true
exp_uni_mi	0.03665375140248242	4.7459725488702547e-7	1	true
exp_uni_mi	0.02330123536326822	5.931317704847098e-5	1	true
exp_uni_mi	0.023052282967998782	6.496292755558229e-5	1	true
exp_uni_mi	0.022007760963460864	9.521107734869327e-5	1	true
exp_uni_mi	0.030866529516644592	3.8070195996482853e-6	1	true
exp_uni_mi	0.00012106058971795245	0.7722467844047455	1	true
exp_uni_mi	0.008572502179418644	0.014866910863161541	1	true
exp_uni_mi	0.0009966001465253013	0.4062843724862705	1	true
exp_uni_mi	0.035717188542120276	6.642024391558446e-7	1	true
exp_uni_mi	0.004941773113932489	0.06442244650625958	1	true
exp_uni_mi	0.06110280167939912	7.897047673449372e-11	1	true
exp_uni_mi	0.015319354667186239	0.0011302691436053875	1	true
exp_uni_mi	-0.0003480730246973201	0.6235809583013783	1	true
exp_uni_mi	0.0034737422843147776	0.12103861525413447	1	true
exp_uni_mi	0.02345987908881297	5.5973378960562814e-5	1	true
exp_uni_mi	-0.005519959761712148	0.05064991624505534	1	true
exp_uni_mi	0.03206111290043406	2.4743437248741325e-6	1	true
exp_uni_mi	0.00010324793638038038	0.7892406609283347	1	true
exp_uni_mi	-0.028877525681507078	7.812361439880574e-6	1	true
exp_uni_mi	0.03804022709206085	2.887033357844865e-7	1	true
exp_uni_mi	-0.0002463300199837858	0.6797029826270802	1	true
exp_uni_mi	0.007386735305070528	0.02376610733945402	1	true
exp_uni_mi	0.008268219060411629	0.016757380733853112	1	true
exp_uni_mi	-2.9029155141292173e-7	0.9886917404803992	1	true
exp_uni_mi	0.02911394266808131	7.171832202218454e-6	1	true
exp_uni_mi	0.01689728155347146	0.0006273840328567913	1	true
exp_uni_mi	0.016213929515498687	0.0008091878469715618	1	true
exp_uni_mi	0.0188433713010917	0.0003049711186687411	1	true
exp_uni_mi	0.07111354007410982	2.2990854806663175e-12	1	true
exp_uni_mi	0.027861460557154494	1.1287488862845044e-5	1	true
exp_uni_mi	0.00668463203900218	0.03149520668253424	1	true
exp_uni_mi	0.002646107699365958	0.1759969038212162	1	true
exp_uni_mi	0.004586920359009424	0.07481217950586622	1	true
exp_uni_mi	-0.005574264426263079	0.04952728776598973	1	true
exp_uni_mi	0.0043710048046410556	0.08200370014785116	1	true
exp_uni_mi	0.015942549734737965	0.0008954067369819787	1	true
exp_uni_mi	0.0010153326514466398	0.4019080184861587	1	true
exp_uni_mi	0.0602626227715629	1.0631460394381767e-10	1	true
exp_uni_mi	0.006903094124479947	0.02884331178967639	1	true
exp_uni_mi	0.03222405742836245	2.3332228245726753e-6	1	true
exp_uni_mi	0.05009795499621008	3.9104958477068585e-9	1	true
exp_uni_mi	0.011176498860229582	0.00541864488996616	1	true
exp_uni_mi	0.011937207093108372	0.0040515911386711494	1	true
exp_uni_mi	0.04502307501556368	2.3810620282318333e-8	1	true
exp_uni_mi	0.01708925999685874	0.0005841661716855527	1	true
exp_uni_mi	-0.0012799051276654186	0.3466467719545408	1	true
exp_uni_mi	0.008051837582535947	0.018251242220733783	1	true
exp_uni_mi_nz	-0.021475022105234293	0.27280239384261845	1	true
exp_uni_mi_nz	-0.009830857043609535	0.40679094125850673	1	true
exp_uni_mi_nz	-0.030454272259192734	0.011055926800267341	1	true
exp_uni_mi_nz	0.011731920234914598	0.23935875668478132	1	true
exp_uni_mi_nz	-0.03077439834436103	0.08564715004021302	1	true
exp_uni_mi_nz	0.02279228664540911	0.003092223284425999	1	true
exp_uni_mi_nz	-0.019551345502153136	0.1538821142833069	1	true
exp_uni_mi_nz	0.0004955225849496082	0.8089280486089445	1	true
exp_uni_mi_nz	0.010631707180484737	0.3446485026890621	1	true
exp_uni_mi_nz	-0.014299912677948905	0.17274157015247021	1	true
exp_uni_mi_nz	-2.270565933729922e-6	0.9779645694531696	1	true
exp_uni_mi_nz	-0.010933854980783085	0.3106800952896053	1	true
exp_uni_mi_nz	-0.00010574153145053797	0.895868562923032	1	true
exp_uni_mi_nz	1.9452806831402747e-6	0.9834203091536105	1	true
exp_uni_mi_nz	-0.0030017091501106635	0.3863409827289426	1	true
exp_uni_mi_nz	0.003494805836653352	0.5706943291436598	1	true
exp_uni_mi_nz	0.03371910842193473	0.0235794335025305	1	true
exp_uni_mi_nz	-0.013976404108940311	0.3027124241752733	1	true
exp_uni_mi_nz	0.0007588159132472514	0.6581533820067905	1	true
exp_uni_mi_nz	-0.007898768337769183	0.16505419901547658	1	true
exp_uni_mi_nz	0.0	1.0	0	false
exp_uni_mi_nz	-0.00020845158195181013	0.8550914883107795	1	true
exp_uni_mi_nz	0.00465560334880518	0.551955355693483	1	true
exp_uni_mi_nz	-0.00030656769883670784	0.7586279628615319	1	true
exp_uni_mi_nz	-0.00046207942207860525	0.8184688007668478	1	true
exp_uni_mi_nz	0.0009120044320882925	0.7696794969705385	1	true
exp_uni_mi_nz	0.002787876842847502	0.4620814556122505	1	true
exp_uni_mi_nz	-0.0005997242579469244	0.6477850644145697	1	true
exp_uni_mi_nz	0.003576461457314692	0.5342730295140183	1	true
exp_uni_mi_nz	-0.0007089181288109462	0.7300145800731231	1	true
exp_uni_mi_nz	1.1337712319815024e-5	0.9562049176722769	1	true
exp_uni_mi_nz	-0.005788431577270654	0.21294367799112426	1	true
exp_uni_mi_nz	-0.007846776482042591	0.38543642118117805	1	true
exp_uni_mi_nz	-0.000654786017886848	0.8061165174275924	1	true
exp_uni_mi_nz	-5.102308263079149e-5	0.9490583475843515	1	true
exp_uni_mi_nz	0.0	1.0	1	true
exp_uni_mi_nz	-0.0012715607976016563	0.6960750066240662	1	true
exp_uni_mi_nz	0.028978618031730047	0.006455567117594121	1	true
exp_uni_mi_nz	0.0014211969027927984	0.6420878707059129	1	true
exp_uni_mi_nz	-0.022062121959458593	0.019330065842752747	1	true
exp_uni_mi_nz	-0.036973131447797444	0.020158948924391473	1	true
exp_uni_mi_nz	-0.0029749323647596214	0.5854582207987026	1	true
exp_uni_mi_nz	-0.009413714380403544	0.09848069262084169	1	true
exp_uni_mi_nz	-0.0011109558762691096	0.6583547996259492	1	true
exp_uni_mi_nz	0.0006749977666700913	0.7796149094482575	1	true
exp_uni_mi_nz	-0.020228545999997558	0.11922856803392792	1	true
exp_uni_mi_nz	0.004181450096760502	0.2533724258786989	1	true
exp_uni_mi_nz	-0.0016334192270084514	0.6553049468817591	1	true
exp_uni_mi_nz	0.017387124833221083	0.09742699265156839	1	true
exp_uni_fz	-0.14517769425573282	0.00677184233961169	0	true
exp_uni_fz	-0.12536808374274133	0.019588931030603426	0	true
exp_uni_fz	0.15152598921918628	0.004682823694564127	0	true
exp_uni_fz	0.019517293623188596	0.717717301144521	0	true
exp_uni_fz	-0.005297394823305385	0.9218449129738744	0	true
exp_uni_fz	-0.0034781650172207037	0.9486383713813415	0	true
exp_uni_fz	0.05517154546761604	0.30638824885040805	0	true
exp_uni_fz	-0.08316260573027531	0.1226468515305277	0	true
exp_uni_fz	-0.00036747698993379893	0.9945698235896551	0	true
exp_uni_fz	-0.10865129561911506	0.04336005506694478	0	true
exp_uni_fz	-0.010588461909730158	0.8445253310561258	0	true
exp_uni_fz	-0.039786517173322364	0.46097289440375067	0	true
exp_uni_fz	0.10782768228597518	0.04498543436179573	0	true
exp_uni_fz	-0.03938265851051115	0.4655402730658161	0	true
exp_uni_fz	-0.18790597781607907	0.00042843911775877205	0	true
exp_uni_fz	-0.04480173344242429	0.40637171227311386	0	true
exp_uni_fz	0.14935293259992766	0.005321464117340604	0	true
exp_uni_fz	-0.1123450798529132	0.03666449010559563	0	true
exp_uni_fz	-0.05896652331129571	0.2742429334217861	0	true
exp_uni_fz	-0.18079618588568064	0.0007102071968488115	0	true
exp_uni_fz	-0.18416975752773979	0.0005600886496505489	0	true
exp_uni_fz	0.022171240418980356	0.6813039524525768	0	true
exp_uni_fz	-0.09486312894781578	0.07803746029450431	0	true
exp_uni_fz	-0.182145852950466	0.0006461704050245584	0	true
exp_uni_fz	-0.04971728583867794	0.35677023326756085	0	true
exp_uni_fz	-0.08828297235830537	0.10115262130908732	0	true
exp_uni_fz	-0.054511004658096905	0.31222722619977694	0	true
exp_uni_fz	-0.10421861902384479	0.05272213743729862	0	true
exp_uni_fz	-0.004140160004843343	0.9388802836791426	0	true
exp_uni_fz	-0.006608634186422339	0.9025858815517713	0	true
exp_uni_fz	-0.00744642546049284	0.8903097462560381	0	true
exp_uni_fz	-0.068006252539392	0.20715018990912615	0	true
exp_uni_fz	-0.0695456999269231	0.19702046801341833	0	true
exp_uni_fz	-0.06952936649765129	0.19712601633132706	0	true
exp_uni_fz	-0.06770394851677429	0.20918274791446897	0	true
exp_uni_fz	-0.12102409582781422	0.024294610845830383	0	true
exp_uni_fz	-0.05156675513376308	0.3391341078369011	0	true
exp_uni_fz	-0.06370607673551872	0.23742278194160377	0	true
exp_uni_fz	-0.11031333076512093	0.040229979143548276	0	true
exp_uni_fz	0.19499468995254318	0.0002539957528280733	0	true
exp_uni_fz	-0.09420322790049593	0.08014303983103029	0	true
exp_uni_fz	-0.011634843744095822	0.8293852722479162	0	true
exp_uni_fz	-0.05639514351164011	0.29576327588080925	0	true
exp_uni_fz	0.005747011650742003	0.9152353864729041	0	true
exp_uni_fz	-0.019444156258292423	0.7187303221811755	0	true
exp_uni_fz	0.08092662780548589	0.13307982199400697	0	true
exp_uni_fz	-0.10300221143128872	0.05556686683840466	0	true
exp_uni_fz	-0.15119376444576413	0.004775763992102074	0	true
exp_uni_fz	-0.0501831697204768	0.35227443723088525	0	true
exp_condZ3_mi_nz	0.0	1.0	0	false
exp_condZ1_fz	-0.13485194932350858	0.011977741390295364	0	true
exp_condZ3_mi	-0.014073193092776139	0.08298837674054324	5	true
exp_uni_fz_nz	-0.4174264484667523	0.026225460974008383	0	true
exp_uni_fz_nz	-0.28108759096983194	0.10224762085463926	0	true
exp_uni_fz_nz	-0.10864275178865991	0.26830577859571136	0	true
exp_uni_fz_nz	0.03721678634248714	0.7805260100128106	0	true
exp_uni_fz_nz	-0.08577089355913897	0.5640841402540078	0	true
exp_uni_fz_nz	0.20920805477603643	0.0035089301055264564	0	true
exp_uni_fz_nz	-0.055311258040693013	0.6983310395227331	0	true
exp_uni_fz_nz	0.013690908554514634	0.9183917354675555	0	true
exp_uni_fz_nz	0.09823307301081258	0.5382575084183825	0	true
exp_uni_fz_nz	-0.2054258748535495	0.10082397014236624	0	true
exp_uni_fz_nz	-0.0013216929343099891	0.9864545950824573	0	true
exp_uni_fz_nz	-0.16026714331871042	0.2835687393074964	0	true
exp_uni_fz_nz	-0.14219576532192107	0.20608384161304943	0	true
exp_uni_fz_nz	0.12832381374081278	0.17992908957732903	0	true
exp_uni_fz_nz	-0.1313154187622195	0.14460626969746843	0	true
exp_uni_fz_nz	-0.06502617313477141	0.6693737437301627	0	true
exp_uni_fz_nz	0.21130222384637043	0.06680560055028012	0	true
exp_uni_fz_nz	-0.0837299443193867	0.6195307432176838	0	true
exp_uni_fz_nz	0.09964111529794265	0.2617806136280858	0	true
exp_uni_fz_nz	-0.1882116204585127	0.03771687999563765	0	true
exp_uni_fz_nz	-0.42934543829015787	0.06630211767407959	0	true
exp_uni_fz_nz	-0.1887173011720694	0.09372299914271348	0	true
exp_uni_fz_nz	0.03846095230920534	0.819918068481526	0	true
exp_uni_fz_nz	-0.14891467497072294	0.06524096646764334	0	true
exp_uni_fz_nz	-0.07993419299591784	0.5560962048324166	0	true
exp_uni_fz_nz	-0.02021384954884321	0.8933220955348123	0	true
exp_uni_fz_nz	0.0007800424499707911	0.9939658212085398	0	true
exp_uni_fz_nz	-0.014757916656919376	0.8469598967223138	0	true
exp_uni_fz_nz	-0.08043755571501951	0.5648296093680875	0	true
exp_uni_fz_nz	-0.16528508658335062	0.13326800592056684	0	true
exp_uni_fz_nz	0.15533080927972126	0.07418069685216865	0	true
exp_uni_fz_nz	-0.13719892989318566	0.11404292216852818	0	true
exp_uni_fz_nz	-0.19604165165676815	0.18275040972598058	0	true
exp_uni_fz_nz	-0.1938188258310108	0.1980111898114465	0	true
exp_uni_fz_nz	-0.07951289146131844	0.6279025057544988	0	true
exp_uni_fz_nz	0.061221748839424946	0.74132286256108	0	true
exp_uni_fz_nz	-0.07648030245905153	0.5628973061670327	0	true
exp_uni_fz_nz	0.26281623528509307	0.002621277951573173	0	true
exp_uni_fz_nz	-0.006280909984914979	0.9572022086739403	0	true
exp_uni_fz_nz	-0.1365128494492382	0.13237085866153653	0	true
exp_uni_fz_nz	-0.43765661633584385	8.612495281749015e-5	0	true
exp_uni_fz_nz	-0.1465710998053305	0.31147555197907834	0	true
exp_uni_fz_nz	-0.3243499034865977	6.075216313259153e-5	0	true
exp_uni_fz_nz	0.15913873782310498	0.1389362634002838	0	true
exp_uni_fz_nz	0.06725626739272911	0.6173990525129001	0	true
exp_uni_fz_nz	-0.16640037299705465	0.20476687688088752	0	true
exp_uni_fz_nz	0.17404828727312524	0.029630041697513328	0	true
exp_uni_fz_nz	-0.1499144564217741	0.24998541788514322	0	true
exp_uni_fz_nz	-0.1099878176255904	0.33567481273015093	0	true
